packages/@react-aria/virtualizer/src/Virtualizer.tsx

```diff
@@ -145,6 +145,17 @@ export function useVirtualizer<T extends object, V, W>(props: VirtualizerOptions
     // element, and we aren't moving focus to the collection from within (see below).
     if (e.target === ref.current && !isFocusWithin.current) {
       virtualizer.scrollToItem(focusedKey, {duration: 0});
+    } else if (e.target !== ref.current && focusedKey != null && !isFocusWithin.current) {
+      let focusedView = virtualizer.getView(focusedKey);
+      // Otherwise if we are re-entering the collection and there is a previously focused key that is out of view, scroll it into view
+      // If it is in view already, preserve the current scroll position via scrollTo. This is to prevent the shift-tabbing into the
+      // collection from scrolling the focused key out of view due to useSelectableCollection focusing the last element in the table when tabbing out of the collection
+      if (!focusedView) {
+        virtualizer.scrollToItem(focusedKey, {duration: 0});
+      } else {
+        let rect = virtualizer.getVisibleRect();
+        virtualizer.scrollTo(new Point(rect.x, rect.y), 0);
+      }
```

Comment:
Fixes "When I shift tab back into the table, it does another page down"

Reply:
Is the root cause here that the focused key actually changes when tabbing out of the table due to focusing the last focusable item? Should we instead prevent this from happening?

Reply:
So the focused key actually doesn't change when tabbing out of the table because the last focusable item is actually a checkbox for a row that is out of view (useGridCell actually doesn't update the focusedKey in this case because we are in keyboard modality and useSelectableItem doesn't either because e.target isn't the cell itself).

The root cause of the weird scrolling behavior from shift tabbing into the table is because the browser is moving focus to the last focusable element in the table which is the out of view checkbox, causing the scroll...

Reply:
Tried a bunch of stuff but this still feels like the best solution since it fixes the following two cases:
- stopping the table from scrolling to the last checkbox when shift tabbing back into the table
- properly scrolling a item back into view when tabbing into the table

The second one can be reproduced by doing the following:
1. focus a cell in the table by clicking/tabbing into the table
2. using your mouse/touchpad, scroll the cell just out of view by scrolling down a little bit (make sure not to scroll so far that the cell is removed from the DOM)
3. click on an element outside of the table
4. tab back into the table

Without the above code, tabbing back into the table would focus the first focusable item, aka the column's select all checkbox which results in the focused cell in the table not being scrolled back into view

Reply:
An alternative would be to apply `excludeFromTabOrder` to the props returned by `useTableSelectionCheckbox` and `useTableSelectAllCheckbox` so that tabbing/shift-tabbing into the table moves focus to the previously focused cell that has tabindex=0.

Reply:
A possible solution that could replace the Virtualizer change here is perhaps something like this?
```
  useEffect(() => {
    let onKeyDown = (e) => {
      // The consts here can be removed if we export them from FocusScope
      const focusableElements = [
        'input:not([disabled]):not([type=hidden])',
        'select:not([disabled])',
        'textarea:not([disabled])',
        'button:not([disabled])',
        'a[href]',
        'area[href]',
        'summary',
        'iframe',
        'object',
        'embed',
        'audio[controls]',
        'video[controls]',
        '[contenteditable]'
      ];

      focusableElements.push('[tabindex]:not([tabindex="-1"]):not([disabled])');
      const TABBABLE_ELEMENT_SELECTOR = focusableElements.join(':not([hidden]):not([tabindex="-1"]),');

      if (e.key === 'Tab' && !ref.current.contains(e.target as HTMLElement)) {
        let tabbableElements = Array.from(document.querySelectorAll(TABBABLE_ELEMENT_SELECTOR));
        let index = tabbableElements.findIndex(node => node === e.target)
        let nodeToFocus;
        if (e.shiftKey) {
          nodeToFocus = tabbableElements[index - 1];
        } else {
          nodeToFocus = tabbableElements[index + 1];
        }

        if (ref.current.contains(nodeToFocus) && manager.focusedKey) {
          e.preventDefault();
          manager.setFocused(true);
          focusSafely(ref.current);
        }
      }
    }

    window.addEventListener('keydown', onKeyDown, true);
    return () => {
      window.removeEventListener('keydown', onKeyDown, true);
    }
  }, [])
```
Essentially we capture all keypresses that happen outside of the TableView. If we detect that the key press is Tab/Shift + Tab we then determine if focus would've shifted into the TableView and if so, prevent default behavior of the Tab key down so the browser doesn't shift the scroll position and then focus the TableView itself. The TableView then handles focusing the previously focused key itself via its existing useVirtualizer code (if selectedKey is out of view) or the useSelectableItem code (if selectedKey is already in view)

Reply:
Another idea is to mimic some of the code in `useSelectableList` used for scrolling the focusedKey into view: https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/selection/src/useSelectableList.ts#L114-L121. We could add similiar `scrollIntoView` in a capturing onFocus listener that detects when the focus is moved into the Table from outside. If so, then we bring the focused item into view but only if it isn't in view already and if `isVirtualized` is false. For the virtualized case, we keep the changes already made in this PR.